### PR TITLE
bgp_communities - Update 'aann' option, Fix issues in merged state

### PR DESCRIPTION
**Pull request by ArunSaravananBalachandran**
##### SUMMARY

Incorporate the below changes in sonic_bgp_communites module:
- Change 'aann' as a suboption of 'members' and update it to support 'list of strings'. 
- Fix issue in merged state for standard community-list when community options are specified as 'False'.
- Update default value behavior for 'type' and 'match' to be in-line with 'permit'.
- Update documentation, UT and regression test cases.
- Resolve ansible-lint reported failures.

##### GitHub Issues
List the GitHub issues impacted by this PR. If no Github issues are affected, please indicate this with "N/A".

| GitHub Issue # |
| -------------- |
| N/A |

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

- sonic_bgp_communities

##### Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.
- [x] I have provided a summary for this PR in valid "fragment" file format in the "changelogs/fragments" directory of this repository branch. Reference : [Ansible Change Log Document](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

##### How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

Regression report: [bgp_communities_regression_report-2024-09-26-11-18-07.pdf](https://github.com/user-attachments/files/17143422/bgp_communities_regression_report-2024-09-26-11-18-07.pdf)

### CODE REVIEW

**Review by kerry-meyer — CHANGES_REQUESTED** [review]
Thank you for providing these changes to provide more complete support in the enterprise_sonic Ansible resource module collection for the BGP Communities functionality currently implemented in SONiC.

The proposed changes, updated test cases, and test results look good.

I am posting only a few minor comments/questions.

File: plugins/module_utils/network/sonic/config/bgp_communities/bgp_communities.py [hunk 1]
```diff
@@ -238,16 +219,45 @@ def _state_overridden(self, want, have):
...
-        requests = self.get_modify_bgp_community_requests(commands, have, "merged")
-        if commands and len(requests) > 0:
+        requests = []
+        commands = get_diff(want, have)
+
+        for conf in commands:
+            have_conf = search_obj_in_list(conf['name'], have, 'name')
+            if have_conf:
+                del_attrs = []
+                for attr in ('type', 'permit', 'match'):
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/bgp_communities/bgp_communities.py:238]: 'want' is updated in `validate_and_normalize_config`, but 'diff' will not have the required fields is they are the same in 'want' and 'have'. Hence, they are being updated here.

Have added a comment on the same.

File: plugins/module_utils/network/sonic/config/bgp_communities/bgp_communities.py [hunk 2]
```diff
@@ -465,78 +432,136 @@ def get_replaced_overridden_config(self, want, have, cur_state):
...
+        else:
+            updated_want = want
+
+        for conf in updated_want:
+            # Empty values for member suboption is supported.
+            # Hence, remove_empties is not used for deleted state.
+            delete_name_only = False
+            if state == 'deleted':
+                for key, value in list(conf.items()):
+                    if value is None:
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/bgp_communities/bgp_communities.py:532]: Yes, empty options for 'regex' and 'aann' are supported (not members, due to the default ansible behavior).
Have updated the comment to specify the supported behavior.

File: plugins/module_utils/network/sonic/config/bgp_communities/bgp_communities.py [hunk 3]
```diff
@@ -322,140 +332,97 @@ def get_delete_bgp_communities(self, commands, have, is_delete_all):
...
+                            requests.append(self.get_delete_single_bgp_community_request(name))
                             break
 
-                        if cmd_type == "standard":
+                        member_type = 'aann' if cmd['type'] == 'standard' else 'regex'
+                        if cmd['type'] == "standard":
+                            have_attr = False
                             for attr in self.standard_communities_map:
-                                if cmd.get(attr, None) and item[attr] and cmd[attr] == item[attr]:
+                                if cmd.get(attr) and item.get(attr) and cmd[attr] == item[attr]:
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/bgp_communities/bgp_communities.py:355]: The variable name has been updated as suggested.

File: plugins/modules/sonic_bgp_communities.py [hunk 4]
```diff
@@ -57,22 +57,17 @@
         type: str
         description:
         - Whether it is a standard or expanded community-list entry.
+        - Default value while adding a new community-list is C(standard).
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/modules/sonic_bgp_communities.py]: As suggested, have updated the description for similar options (match, permit).

[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/modules/sonic_bgp_communities.py]: Thanks for making these changes.

Resolved.

File: plugins/module_utils/network/sonic/config/bgp_communities/bgp_communities.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/bgp_communities/bgp_communities.py:532]: Thanks for clarifying this.

Resolved.

File: plugins/module_utils/network/sonic/config/bgp_communities/bgp_communities.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/bgp_communities/bgp_communities.py:238]: Thanks for explaining and for providing the explanation in a comment for future readers and maintainers.

Resolved.

File: plugins/module_utils/network/sonic/config/bgp_communities/bgp_communities.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/bgp_communities/bgp_communities.py:355]: Thanks for making this change.

Resolved.

**Review by kerry-meyer — APPROVED** [review]
Thanks for addressing the posted comments.

All current proposed code changes and comments, test cases, and test results look good.

Approved.

**Pull request merged**